src/CRISP.BackendChallenge/Controllers/EmployeeController.cs

```diff
@@ -7,33 +7,266 @@
+        {
+            _logger.LogError(":: Error occurred while getting employee ID: {Id} {EMessage}", id, e.Message);
+            return CreateInternalServerError();
+        }
+    }
+    
+    
+    [HttpPost]
+    [Route("createEmployee")]
+    public IActionResult AddEmployee([FromBody] EmployeeRequest employeeRequest)
+    {
+        _logger.LogDebug(":: Performing {MethodName}", nameof(AddEmployee));
+        
+        if (string.IsNullOrWhiteSpace(employeeRequest.Name))
+        {
+            _logger.LogError(":: Invalid EmployeeRequest provided");
+            return CreateBadRequest("Creating an employee", "Name field required");
+        }
+        if (employeeRequest.Department == null)
+        {
+            _logger.LogError(":: Invalid EmployeeRequest provided");
+            return CreateBadRequest("Creating an employee", "Department field required");
+        }
+        
+        try
+        {
+            
+            var employee = new Employee
+            {
+                Name = employeeRequest.Name,
+                Department = (Department)employeeRequest.Department,
+                StartDate = employeeRequest.StartDate,
+                EndDate = employeeRequest.EndDate
+            };
+
+            _employeeRepository.Add(employee);
+            _employeeRepository.Save();
+
+            if (employeeRequest.LoginDates != null || employeeRequest.LoginDates!.Any())
+            {
+                foreach (var login in employeeRequest.LoginDates!.Select(l => new Login()
+                         {
+                             EmployeeId = employee.Id,
+                             LoginDate = l
+                         }))
+                {
+                    _loginRepository.Add(login);
+                }
+                _loginRepository.Save();
+            }
+            
+            return Ok(ToEmployeeResponse(employee));
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(":: Error occurred while creating employee: {EMessage}", e.Message);
+            return CreateInternalServerError();
+        }
+    }
+
+    
+    [HttpGet]
+    [Route("search")]
+    public IActionResult SearchEmployees(int? id, string? name, int? department)
+    {
+        _logger.LogDebug(":: Performing {MethodName}", nameof(SearchEmployees));
+
+        try
+        {
+            if (id == null && string.IsNullOrWhiteSpace(name) && department == null)
+            {
+                var employees = _employeeRepository.GetAll();
+
+                var res = employees.Select(ToEmployeeResponse);
+
+                return Ok(res);
+            }
+            
+            var result = _employeeRepository.Query().Include(e => e.Logins)
+                .Where(x => (id.HasValue && x.Id == id) || (!string.IsNullOrEmpty(name) && x.Name == name) ||
+                            (department.HasValue && x.Department == (Department)department))
+                .Select(x => ToEmployeeResponse(x)).ToList();
```

Comment:
What would be slightly better is to conditionally build your query based on the inputs:

```cs
var baseQuery = _employeeRepository.Query().Include(e => e.Logins); // baseQuery is IQueryable<Employee>

// Realistically you wouldn't support the primary key (ID) as a parameter on your query endpoint 
// - you would use an endpoint like `GET api/Employee/{id}`

if (!string.IsNullOrWhitespace(name))
  baseQuery = baseQuery.Where(x => x.Name == name);

if (department.HasValue)
  baseQuery = baseQuery.Where(x => x.Department == department);

var result = baseQuery.ToList(); // Execute the query

..
```

Reply:
Removed ID as a parameter, and cleaned up how the controller flows